lightning/src/ln/outbound_payment.rs

```diff
@@ -553,68 +569,167 @@ impl OutboundPayments {
 			!pmt.is_auto_retryable_now() && pmt.remaining_parts() == 0 && !pmt.is_fulfilled())
 	}
 
-	/// Will return `Ok(())` iff at least one HTLC is sent for the payment.
-	fn pay_internal<R: Deref, NS: Deref, ES: Deref, IH, SP, L: Deref>(
-		&self, payment_id: PaymentId, payment_hash: PaymentHash,
-		initial_send_info: Option<(&Option<PaymentSecret>, Option<PaymentPreimage>, Retry)>,
-		route_params: RouteParameters, router: &R, first_hops: Vec<ChannelDetails>,
-		inflight_htlcs: &IH, entropy_source: &ES, node_signer: &NS, best_block_height: u32,
-		logger: &L, pending_events: &Mutex<Vec<events::Event>>, send_payment_along_path: &SP,
-	) -> Result<(), PaymentSendFailure>
+	/// Errors immediately on [`RetryableSendFailure`] error conditions. Otherwise, further errors may
+	/// be surfaced asynchronously via [`Event::PaymentPathFailed`] and [`Event::PaymentFailed`].
+	///
+	/// [`Event::PaymentPathFailed`]: crate::util::events::Event::PaymentPathFailed
+	/// [`Event::PaymentFailed`]: crate::util::events::Event::PaymentFailed
+	fn send_payment_internal<R: Deref, NS: Deref, ES: Deref, IH, SP, L: Deref>(
+		&self, payment_id: PaymentId, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>,
+		keysend_preimage: Option<PaymentPreimage>, retry_strategy: Retry, route_params: RouteParameters,
```

Comment:
Parameters are a little inconsistent here. Should we pass `payment_secret` by value?

Reply:
It's preexisting and I think it would be better to fix it all in one go, since `ChannelManager::send_payment` and friends also need this fix